File: software/rf_devel.py

import matplotlib.pyplot as plt
import numpy as np
import pandas
import pdb
import crc8

#SOURCE_FILE_NAME = 'transmitted_data.csv'
#SOURCE_FILE_NAME = 'received_data.csv'
SOURCE_FILE_NAME = 'generated_packet.csv'


df = pandas.read_csv(SOURCE_FILE_NAME, header=None, names=['t', 'val', 'notes'])
df['t'] = df['t'] * 1e6  # convert to us
df.sort_values('t', inplace=True)

#print(df)

def sample_rf_(t, df_):
	# In the real implementation, this will just be a periodic sample of the bus.
	# Here, it finds the current sample of the "bus" from the data file based on
	# current time increment.
	last_edge_row = df_[df_['t'].lt(t)].tail(1)
	if len(last_edge_row) > 0:
		#print('currently {} - last edge at {}'.format(t, last_edge_row['t'].values[0]))
		return last_edge_row['val'].values[0]
	else:
		return 0

class SampleBuffer:
	def __init__(self, length):
		self.vals = np.zeros(length)
		self.len = length
		self.cnt = 0

	def put(self, val):
		self.vals[self.cnt] = val;
		self.cnt += 1
		if self.cnt == self.len:
			pdb.set_trace()
			raise Exception('Overflowed sample buffer')

	def is_full(self):
		return self.cnt == (self.len - 1)

	def peek(self):
		if(self.cnt > 0):
			return self.vals[0]
		else:
			return -1

	def flush(self):
		self.vals = np.zeros(self.len)
		self.cnt = 0

	def pop(self):
		if(self.cnt > 0):
			popped = self.vals[0]
			self.vals = np.append(self.vals[1:],0)
			self.cnt -= 1
			return popped
		else:
			return -1



def CheckBitCorrelation(samples, expected_bit):
	# TODO: find a better correlation algorithm. This rewards longer bits. Normalize to bit duration?
	c = 0
	for i in samples:
		if i == expected_bit:
			c += 1
	return c


def ProcessSample(samples):
	sym_len = int(len(samples) / 2)
	one_val = int(sym_len/2) + 1
	assert(sym_len*2 == len(samples)) # shouldn't be receiving an odd number of samples
	sym1 = sum(samples[0:sym_len-1]) >= one_val
	sym2 = sum(samples[sym_len:-1]) >= one_val
	if sym1 and not sym2: # 0b10
		return 0
	if not sym1 and sym2: # 0b01
		return 1
	#pdb.set_trace()
	return -1
	
def calc_crc(data, length):
	return int(crc8.crc8(data.to_bytes(length, 'big')).digest().hex(), 16)

kLoopDuration_us = 10.0
kMinBitDuration_us = 40.0
kMaxBitDuration_us = 60.0
kMinCorrelation = int(kMaxBitDuration_us/kLoopDuration_us*0.8)
kPreambleCorrelationLen = 6
kMaxSampleCnt = int(kMaxBitDuration_us/kLoopDuration_us * 5) # Need to store up to five bit periods of data
sb = SampleBuffer(kMaxSampleCnt+1)
kPreambleLen = 20
kLengthLen = 4
state_ = 'kSync'
samples_remaining_ = 0
bit_offset = 0
bit_dur = 1
bit_cnt = 0
data = 0
length_bytes = 0
def DoRfReceive():	
	global state_, sb, samples_remaining_, bit_offset, bit_dur, bit_cnt, data, length_bytes
	#TODO Make sure interrupts are disabled when we're doing this for real to avoid conflicts on sample buffer
	if state_ == 'kSync':
		while (sb.peek()) == 0:
			sb.pop() # wait for a rising edge
		
		if sb.is_full(): 
			# Find bit offset and duration that maximizes correlation with [1,0]
			bit_correlation = 0
			bit_offset = 0
			bit_dur = 1
			for o in range(0,(int(kMaxBitDuration_us/kLoopDuration_us)*2)+1): # Offsets from 0 up to two bit periods
				for d in range(int(kMinBitDuration_us/kLoopDuration_us), int(kMaxBitDuration_us/kLoopDuration_us)+1): # Allow durations within valid range
					correlation = 0
					for b in range(0,kPreambleCorrelationLen):
						correlation += CheckBitCorrelation(sb.vals[o+(d*b):o+(d*(b+1))], (b+1)%2)
					#print('Correlation {} for offset {} and duration {}'.format(correlation, o, d))
					if correlation >= bit_correlation: # Better than last best
						bit_correlation = correlation
						bit_offset = o
						bit_dur = d

			if bit_correlation > (kMinCorrelation * kPreambleCorrelationLen):
				print('Found preamble sequence with best offset {} and duration {} and correlation {}'.format(bit_offset, bit_dur, bit_correlation))
				samples_remaining_ = bit_dur - ((kMaxSampleCnt - bit_offset) % bit_dur) - 1
				sb.flush() # Discard samples in buffer, as we assume we've found a preamble
				state_ = 'kAlign'
			else:
				sb.pop() # drop edge and keep trying
			
	elif state_ == 'kAlign':
		#print('Aligning by receiving {} more samples'.format(samples_remaining_))
		if samples_remaining_ == 0:
			sb.flush() # TODO ideally we'd check this bit against an expected sample
			bit_cnt = 0
			state_ = 'kFinishPreamble'
		samples_remaining_ -= 1

	elif state_ == 'kFinishPreamble':
		if sb.cnt == bit_dur * 2:
			bit = ProcessSample(sb.vals[0:bit_dur*2])
			if bit == -1: # invalid Manchester encoding received, abandon packet
				print('detected invalid packet, starting over')
				state_ = 'kSync'
			if bit == 0: # Received stop bit.
				print('Detected Start bit')
				bit_cnt = 0
				length_bytes = 0
				state_ = 'kLength'
			else:
				bit_cnt += 1
				if bit_cnt > kPreambleLen:
					print('Preamble too long, restarting')
					state_ = 'kSync'
			sb.flush()
		# TODO Periodically re-sync on edges? Depending on how well aligned we are this might not be necessary. Might also accidentally sync to noise.
	
	elif state_ == 'kLength':
		if sb.cnt == bit_dur * 2:
			bit_cnt += 1
			bit = ProcessSample(sb.vals[0:bit_dur*2])
			if bit == -1: # invalid Manchester encoding received, abandon packet
				print('detected invalid Manchester Encoding in length, starting over')
				state_ = 'kSync'
				sb.flush()
				return
			length_bytes = (length_bytes << 1) + bit
			if bit_cnt == kLengthLen:
				print('Received Length {}'.format(length_bytes))
				if length_bytes < 1 or length_bytes > 15:
					print('Detected invalid length, starting over')
					state_ = 'kSync'
					return
				bit_cnt = 0
				data = 0
				state_ = 'kData'
			sb.flush()

	elif state_ == 'kData':
		if sb.cnt == bit_dur * 2:
			bit_cnt += 1
			bit = ProcessSample(sb.vals[0:bit_dur*2])
			if bit == -1: # invalid Manchester encoding received, abandon packet
				print('detected invalid manchester encoding in data, starting over')
				state_ = 'kSync'
				sb.flush()
				return
			data = (data << 1) + bit
			#print("{} Got data bit 0b{:b}".format(bit_cnt, bit))
			if bit_cnt == (length_bytes*8 + 8): # Received all data and crc bits
				# Check crc:
				crc = data & 0xFF
				data = data >> 8
				#print("{:0X}, {}, {:0X}, {:0X}".format(data, length_bytes, data, crc))
				if calc_crc(data, length_bytes) != crc:
					print('*******BAD CRC {} **********'.format(crc))
				# Process packet:
				print('***************  Data received: 0x{:X}'.format(data))
				# Reset state machine
				state_ = 'kSync'
			sb.flush()


for t in np.arange(min(df['t']), max(df['t']+2*kMaxBitDuration_us), kLoopDuration_us):
	sb.put(sample_rf_(t, df))
	#print(t)
	DoRfReceive()

#pdb.set_trace()

File: software/rf_devel.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "crc.h"
#include "bitarray.h"

//#define TEST

#define INTERRUPT_PeripheralInterruptEnable() while(0)
#define INTERRUPT_PeripheralInterruptDisable() while(0)

#define kInputFile "generated_packet.csv"

#define kMinSymbolDuration 4
#define kMaxSymbolDuration 6
#define kPreambleCorrelationLen_symbols 20
#define kMinCorrelation (10 * (kPreambleCorrelationLen_symbols * 90 / 100)) // Need at least 90% of bits

// Protocol definitions
#define kMaxPreambleLen_bits 20
#define kLengthLen_bits 4
#define kMaxDataLen_bytes 16 

static BitArray_t sample_buffer_;

const char* getfield(char* line, int num){
    const char* tok;
    char* tmp = strdup(line);
    for (tok = strtok(tmp, ",");
            tok && *tok;
            tok = strtok(NULL, ",\n")){
        if (!--num) {
        	free(tmp);
            return tok;
        }
    }
    free(tmp);
    return NULL;
}

int8_t test_sample_rf(uint32_t time_us, FILE* stream){
	char line[64];
    uint32_t edge_time_us, last_edge_time_us = 0;
    int8_t sample = -1;
    rewind(stream);
    while (fgets(line, 64, stream)){
        last_edge_time_us = edge_time_us;
        edge_time_us = (uint32_t)(atof(getfield(line, 1)) * (double)(1e6));
        if(time_us < edge_time_us) {
    		sample = atoi(getfield(line, 2));
        	break;
        }
    }
    return sample;
}

uint16_t CheckPreambleCorrelation(BitArray_t *p_samples, uint8_t *p_duration, uint8_t *p_phase){
    // Find symbol phase offset and duration that maximizes correlation with header
    uint16_t best_correlation, correlation = 0;
    uint8_t expected_symbol  = 0;
    for (uint8_t duration = kMinSymbolDuration; duration <= kMaxSymbolDuration; duration++) { // Symbol duration within valid range
        for (uint8_t phase = 0; phase < kMaxSymbolDuration; phase++) { // phase offsets from 0 up to 2 expected symbol periods
            correlation = 0;
            expected_symbol = 0;
            bit_t sample;
            for (uint8_t symbol_cnt = 0; symbol_cnt < kPreambleCorrelationLen_symbols; symbol_cnt++) {
                // Cycle through sample array, using current offset and duration
                for (uint8_t i=0; i < duration; i++) {
                    BitArrayPeek(p_samples, &sample, phase+i+(duration*symbol_cnt));
                    if(sample == expected_symbol) {
                        correlation++;
                    }
                }
                // Switch symbols (header is alternating 01010101 because of manchester encoding)
                expected_symbol = (expected_symbol + 1) % 2; // TODO - make this actually pull from the header definition
            }
            // Normalize to current duration, but multiply by 10 to minimize quantization error.
            // This is so that we don't award better correlation to longer symbols.
            correlation =  (10 * correlation) / duration;

            // If correlation is best yet, update best phase and offsets:
            if (correlation > best_correlation) {
                best_correlation = correlation;
                *p_phase = phase;
                *p_duration = duration;
            }
        }
    }
    return best_correlation;
}

uint8_t ManchesterDecodeBitFromSamples(BitArray_t *p_samples, uint8_t duration) {
    uint8_t symbols[2] = {0};
    bit_t sample;
    for (int i=0; i<2; i++){
        for (int j=0; j<duration; j++){
            BitArrayPop(p_samples, &sample);
            symbols[i] += sample;
        }
    }
    // If symbols are more than half 1, make symbol=1.
    symbols[0] = (symbols[0] > duration/2) ? 1 : 0;
    symbols[1] = (symbols[1] > duration/2) ? 1 : 0;

    // Decode manchester encoding (1 = 0->1 transition, 0 = 1->0 transition)
    if(symbols[0] < symbols[1]) {
        return 1;
    } else if (symbols[0] > symbols[1]) {
        return 0;
    } else {
        printf("Error decoding manchester, received sequence %d%d\r\n", symbols[0], symbols[1]);
        return 0;
    }
}

void debug_dump_buffer(void){
    uint8_t sample;
    printf("sample buffer:\r\n");
    for (int i=0; i<BitArraySize(&sample_buffer_); i++){
        BitArrayPeek(&sample_buffer_, &sample, i);
        printf("%d", sample);
    }
    printf("\r\n");
}

void DoRfReceive(void){
    static enum {kSync, kFinishPreamble, kLength, kData} state = kSync;
    static uint8_t duration, phase;
    static uint8_t statebits_received;
    static uint8_t packet_len_bytes;
    static uint8_t message_buffer[kMaxDataLen_bytes+1] = {0}; // complete data packet + CRC byte
    bit_t sample;
    uint16_t correlation;
    uint8_t bit;

    switch (state){
        case kSync:
            // Detect phase and duration of the received signal clock
            // Wait for a rising edge
            BitArrayPeek(&sample_buffer_, &sample, 0);
            while (sample == 0)
                BitArrayPop(&sample_buffer_, &sample); 
            // Wait until we have at least a worst-case preamble's worth of samples
            if(BitArraySize(&sample_buffer_) > kPreambleCorrelationLen_symbols * kMaxSymbolDuration*2) {
                // Find best correlation for the samples we've received
                correlation = CheckPreambleCorrelation(&sample_buffer_, &duration, &phase);
                printf("correlation: %d\r\n", correlation);
                // See if it's good enough to treat as a real packet
                if (correlation > kMinCorrelation) {
                    printf("Found preamble with phase %d and duration %d\r\n",phase, duration);

                    // Determine how many samples we need to advance to align sample buffer to signal phase
                    printf("Align by receiving %d more samples\r\n", phase);
                    for (int i=0; i< phase; i++){
                        BitArrayPop(&sample_buffer_, &sample);
                    }
                    statebits_received = 0; // Initialize bit counter for preamble reception
                    state = kFinishPreamble;
                } else {
                    // Correlation not good enough, drop edge and try again
                    BitArrayPop(&sample_buffer_, &sample); 
                }
            }
            break;
        case kFinishPreamble:
            // Keep checking for start bit (manchester encoded 0), up to maximum length of preamble
            if(BitArraySize(&sample_buffer_) > duration*2){
                bit = ManchesterDecodeBitFromSamples(&sample_buffer_, duration);
                if (bit == 0){
                    printf("Found start bit after %d preamble bits\r\n", statebits_received);
                    statebits_received = 0;
                    state = kLength;
                }
                if (++statebits_received > kMaxPreambleLen_bits) {
                    printf("Error - received too many preamble bits (received %d), discarding packet\r\n",
                        statebits_received);
                    state = kSync;
                }
                // TODO Periodically re-sync on edges? Depending on how well aligned we are this might not be necessary. Might also accidentally sync to noise.
            }
            break;
        case kLength:
            // Receive length byte
            if (BitArraySize(&sample_buffer_) > duration*2*kLengthLen_bits) {
                packet_len_bytes = 0;
                for (int i=0; i<kLengthLen_bits; i++){
                    bit = ManchesterDecodeBitFromSamples(&sample_buffer_, duration);
                    packet_len_bytes = (packet_len_bytes<<1) + bit;
                }
                printf("Detected packet length %d\r\n", packet_len_bytes);
                statebits_received = 0;
                memset(message_buffer, 0, sizeof(message_buffer));
                state = kData;
            }
            break;
        case kData:
            // Read in data bytes
            if(BitArraySize(&sample_buffer_) > duration*2){
                bit = ManchesterDecodeBitFromSamples(&sample_buffer_, duration);
                uint8_t curr_byte = packet_len_bytes - (statebits_received/8);
                uint8_t curr_bit = 7-(statebits_received % 8);
                message_buffer[curr_byte] += bit<<(curr_bit);
                statebits_received++;
                if(statebits_received == packet_len_bytes*8 + 8){
                    printf("Packet: 0x");
                    for (int i=packet_len_bytes; i>=0; i--){
                        printf("%X", message_buffer[i]);
                    }
                    printf("\r\n");
                    uint8_t calc_crc = crc_init();
                    for (int i=kMaxDataLen_bytes; i>=1; i--){
                        calc_crc = crc_update(calc_crc, &message_buffer[i], 1);
                    }
                    calc_crc = crc_finalize(calc_crc);
                    printf("Caclulated CRC = 0x%X\r\n", calc_crc);                    
                    printf("Received CRC = 0x%X\r\n", message_buffer[0]);

                    if (calc_crc == message_buffer[0]){
                        printf("Received valid packet\r\n");
                    }
                    state = kSync;
                }
            }
            break;
    }
}

int main (void){
	FILE* stream = fopen(kInputFile, "r");
	uint32_t time_us = 0;
	int8_t sample = 0;

	#ifdef TEST
	TestBitArray();
	#endif
	
	while (sample != -1){
		// Get sample from CSV file at sample rate
        sample = test_sample_rf(time_us, stream);
		time_us += 10; // increment by 10us sample interval

        // Put sample into buffer (ISR would normally do this)
        BitArrayPush(&sample_buffer_, sample);

		DoRfReceive();
	}
	fclose(stream);
}

